### PR TITLE
support for gemini testing tool

### PR DESCRIPTION
**Pull request by larisau**

### CODE REVIEW

File: sdcm/cluster_aws.py [hunk 1]
```diff
@@ -417,22 +417,21 @@ def add_nodes(self, count, ec2_user_data='', dc_idx=0, enable_auto_bootstrap=Fal
...
-                              endpoint_snitch=endpoint_snitch,
-                              authenticator=self.params.get('authenticator'),
-                              server_encrypt=self._param_enabled('server_encrypt'),
-                              client_encrypt=self._param_enabled('client_encrypt'),
-                              append_scylla_args=self.params.get('append_scylla_args'))
+            setup_params.update(dict(
+                seed_address=seed_address,
+                broadcast=node.public_ip_address,
+            ))
+
```
Comment (roydahan) [inline sdcm/cluster_aws.py]: @larisau, why are these changes needed?

File: tests/gemini_test.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
+test_duration: 10080
+n_db_nodes: 1
+n_loaders: 1
+n_monitor_nodes: 1
+user_prefix: 'gemini-test'
+failure_post_behavior: keep
+ip_ssh_connections: 'public'
```
Comment (roydahan) [inline tests/gemini_test.yaml]: change to private

File: tests/gemini_test.yaml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+experimental: 'true'
+append_conf: "enable_keyspace_column_family_metrics: true"
```
Comment (roydahan) [inline tests/gemini_test.yaml]: What is it?

File: tests/gemini_test.yaml [hunk 4]
```diff
@@ -0,0 +1,59 @@
+test_duration: 10080
```
Comment (roydahan) [inline tests/gemini_test.yaml]: Is it supposed to run a week?

File: tests/gemini_test.yaml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+n_db_nodes: 1
+n_loaders: 1
+n_monitor_nodes: 1
+user_prefix: 'gemini-test'
```
Comment (roydahan) [inline tests/gemini_test.yaml]: change to "gemini-mixed-test-Xh-Ygb"
X = test time
Y = amount of test data.

File: tests/gemini_test.yaml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        cluster_backend: 'aws'
+        instance_type_loader: 'c4.4xlarge'
+        instance_type_monitor: 't2.small'
+        user_credentials_path: '~/.ssh/scylla-test'
+        us_east_1:
+            region_name: 'us-east-1'
+            security_group_ids: 'sg-5e79983a'
+            subnet_id: 'subnet-c327759a'
+            ami_db_scylla_user: 'centos'
+            ami_id_db_scylla: 'ami-0e2ad213d8babc2da'
```
Comment (roydahan) [inline tests/gemini_test.yaml]: reaplce all AMIs with AMI_ID.

File: tests/gemini_test.yaml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            ami_monitor_user: 'centos'
+
+databases: !mux
+    scylla:
+        db_type: scylla
+        instance_type_db: 'i3.4xlarge'
+    mixed_scylla:
+        db_type: mixed_scylla
+        instance_type_db: 'i3.large'
+        instance_type_db_mixed: 'i3.8xlarge'
```
Comment (roydahan) [inline tests/gemini_test.yaml]: for better understanding which instance is this, I would call it:
"instnace_type_db_test_oracle"

File: tests/gemini_test.yaml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        ami_id_db_scylla_mixed: 'ami-0359657c2fb5faa21'
```
Comment (roydahan) [inline tests/gemini_test.yaml]: same here for the name + AMI_ID_TEST_ORACLE for the value.

File: tests/gemini_test.yaml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        append_scylla_args_mixed: '--enable-cache false'
```
Comment (roydahan) [inline tests/gemini_test.yaml]: same here for the name.

File: tests/gemini_test.yaml [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+append_conf: "enable_keyspace_column_family_metrics: true"
+instance_provision: 'spot_low_price'
+# instance_provision can be one of the following: on_demand|spot_fleet|spot_low_price|spot_duration
+store_results_in_elasticsearch: False
+
+# gemini
+gemini_run: True
+# the below cmd runs about 3 hours
+gemini_cmd: "gemini -d -n 1000000 -c 20 -p 100 -m mixed -f"
+gemini_url: 'https://s3.amazonaws.com/scylla-gemini/Binaries/gemini-0.2'
```
Comment (roydahan) [inline tests/gemini_test.yaml]: can we point it to gemini-latest or something like that?

File: sdcm/cluster.py [hunk 11]
```diff
@@ -1850,30 +1861,40 @@ def node_setup(self, node, verbose=False, db_node_address=None, **kwargs):
...
-
-        cs_exporter_setup = CassandraStressExporterSetup()
-        cs_exporter_setup.install(node)
+            node.remoter.run("echo 'export DB_ADDRESS=%s' >> $HOME/.bashrc" % db_node_address)
 
-        if self.params.get('bench_run', default=False):
-            # go1.7 still not in repo
+        bench_run = self.params.get('bench_run', default=False)
+        gemini_run = self.params.get('gemini_run', default=False)
+        if bench_run or gemini_run:
```
Comment (roydahan) [inline sdcm/cluster.py]: Can we replace these params with checking if gemini_cmd and bench_cmd?
Then our YAMLs don't need 2 params for the same behaviour.

File: tests/gemini_test.yaml (hunk 7 above)
Comment (larisau) [reply tests/gemini_test.yaml]: It's "oracle" just for the gemini test, but what if we want to create 2 or more scylla clusters?

Comment (larisau) [reply tests/gemini_test.yaml]: It's "oracle" just for the gemini test, but what if we want to create 2 or more scylla clusters?

File: tests/gemini_test.yaml (hunk 10 above)
Comment (larisau) [reply tests/gemini_test.yaml]: I can create a copy of the last binary on s3 and call it "latest"

File: tests/gemini_test.yaml (hunk 7 above)
Comment (larisau) [reply tests/gemini_test.yaml]: It's "oracle" just for the gemini test, but what if we want to create 2 or more scylla clusters for another purpose?

File: tests/gemini_test.yaml (hunk 10 above)
Comment (larisau) [reply tests/gemini_test.yaml]: I can create on s3 a copy of the last uploaded binary and call it "latest"

File: sdcm/cluster.py (hunk 11 above)
Comment (larisau) [reply sdcm/cluster.py]: The command can be defined inside a test, so it's not a good way, I think.
In general, we can define something like cassandra-stress: False and install the both: scylla-bench and gemini

File: sdcm/cluster_aws.py (hunk 1 above)
Comment (larisau) [reply sdcm/cluster_aws.py]: append_scylla_args has changed + code enhancement

File: tests/gemini_test.yaml (hunk 7 above)
Comment (larisau) [reply tests/gemini_test.yaml]: I don't like the "oracle" name, it's just for gemini, it's not so clear if we want to create multiple scylla clusters

File: tests/gemini_test.yaml (hunk 10 above)
Comment (larisau) [reply tests/gemini_test.yaml]: I can create on s3 copy of last uploaded binary and call it "gemini-latest"

File: sdcm/tester.py [hunk 12]
```diff
@@ -466,13 +461,24 @@ def create_cluster(db_type='scylla'):
...
 
         db_type = self.params.get('db_type')
         if db_type in ('scylla', 'cassandra'):
             self.db_cluster = create_cluster(db_type)
         elif db_type == 'mixed':
             self.db_cluster = create_cluster('scylla')
             self.cs_db_cluster = create_cluster('cassandra')
+        elif db_type == 'mixed_scylla':
+            self.db_cluster = create_cluster('scylla')
+            self.cs_db_cluster = create_cluster('mixed_scylla')
```
Comment (amoskong) [inline sdcm/tester.py]: So here, we are using scylla cluster to simulate c* cluster?
We may run gemini test by `two scylla clusters` or `one scylla cluster + one C cluster`?

Comment (larisau) [reply sdcm/tester.py]: Yes, currently gemini verifies scylla vs scylla stable version with cache disabled, but actually it can be cassandra or something else - it just has to be stable

File: tests/gemini_test.yaml [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            region_name: 'us-east-1'
+            security_group_ids: 'sg-5e79983a'
+            subnet_id: 'subnet-c327759a'
+            ami_db_scylla_user: 'centos'
+            ami_id_db_scylla: 'AMI-ID'
+            ami_id_loader: 'AMI-ID'
+            ami_loader_user: 'centos'
+            ami_id_db_cassandra: 'AMI-ID'
+            ami_db_cassandra_user: 'centos'
+            ami_id_monitor: 'AMI-ID'
```
[the remark this answers is not in the document]
Comment (larisau) [reply tests/gemini_test.yaml]: removed

File: tests/gemini_test.yaml [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+failure_post_behavior: keep
+ip_ssh_connections: 'private'
+experimental: 'true'
+instance_provision: 'spot_low_price'
+# instance_provision can be one of the following: on_demand|spot_fleet|spot_low_price|spot_duration
+store_results_in_elasticsearch: False
+
+# gemini
+# the below cmd runs about 3 hours
+gemini_cmd: "gemini -d -n 1000000 -c 20 -p 100 -m mixed -f"
```
[the remark this answers is not in the document]
Comment (larisau) [reply tests/gemini_test.yaml]: added

File: tests/gemini_test.yaml [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+databases: !mux
+    scylla:
+        db_type: scylla
+        instance_type_db: 'i3.4xlarge'
+    mixed_scylla:
+        db_type: mixed_scylla
+        instance_type_db: 'i3.large'
+        instance_type_db_oracle: 'i3.8xlarge'
+        ami_id_db_scylla_oracle: 'AMI-ID'
```
[the remark this answers is not in the document]
Comment (larisau) [reply tests/gemini_test.yaml]: done

File: tests/gemini_test.yaml [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+oracle_db_nodes_private_ip: []
+loaders_public_ip: []
+loaders_private_ip: []
+monitor_nodes_public_ip: []
+monitor_nodes_private_ip: []
+
+backends: !mux
+    aws: !mux
+        cluster_backend: 'aws'
+        instance_type_loader: 'c4.4xlarge'
```
Comment (roydahan) [inline tests/gemini_test.yaml]: Why do we need such a strong machine for loader?
Can't it be c4.large? or even c4.2xlarge?

Comment (larisau) [reply tests/gemini_test.yaml]: I think we can use c4.2xlarge

File: tests/gemini_test.yaml [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+databases: !mux
+    scylla:
+        db_type: scylla
+        instance_type_db: 'i3.4xlarge'
+    mixed_scylla:
+        db_type: mixed_scylla
+        instance_type_db: 'i3.large'
+        instance_type_db_oracle: 'i3.8xlarge'
+        ami_id_db_oracle: 'AMI-ID'
```
[the remark this answers is not in the document]
Comment (larisau) [reply tests/gemini_test.yaml]: done

File: sdcm/tester.py [hunk 18]
```diff
@@ -466,13 +461,24 @@ def create_cluster(db_type='scylla'):
                     ec2_ami_id=self.params.get('ami_id_db_cassandra').split(),
                     ec2_ami_username=self.params.get('ami_db_cassandra_user'),
                     **cl_params)
+            elif db_type == 'mixed_scylla':
+                cluster.Setup.mixed_cluster(True)
+                cl_params.update(dict(ec2_instance_type=self.params.get('instance_type_db_oracle'),
```
[the remark this answers is not in the document]
Comment (roydahan) [reply sdcm/tester.py]: "test oracle" it's the term for the data that contains the expected result to determine if the test passed or failed.
It's ok to leave it as is.
This is also the term they used for it in Gemini code.

**Comment by roydahan** [comment]
@larisau, can you please help with this to fix conflicts?
@bentsi, I would like to merge it, so if you have any other comment, please comment ASAP.
I'm not going to wait any longer, it's the third time Larisa fixes conflicts.

**Comment by larisau** [comment]
@roy @bentsi rebased

File: tests/gemini_test.yaml [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+instance_provision: 'spot_low_price'
+# instance_provision can be one of the following: on_demand|spot_fleet|spot_low_price|spot_duration
+store_results_in_elasticsearch: False
+
+# gemini
+# cmd: gemini -d -n [NUM_OF_TEST_ITERATIONS] -c [NUM_OF_THREADS] -p [NUM_OF_PARTITION_KEYS_PER_THREAD] -m mixed -f
+# the below cmd runs about 3 hours
+gemini_cmd: "gemini -d -n 1000000 -c 20 -p 100 -m mixed -f"
+gemini_url: 'https://s3.amazonaws.com/scylla-gemini/Binaries/gemini-latest'
+gemini_static_url: 'https://s3.amazonaws.com/scylla-gemini/Binaries/schema.json'
```
[the remark this answers is not in the document]
Comment (bentsi) [reply tests/gemini_test.yaml]: I meant  `gemini_static_url`... I don't understand what it is used for

File: tests/gemini_test.yaml [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+            ami_monitor_user: 'centos'
+
+databases: !mux
+    scylla:
+        db_type: scylla
+        instance_type_db: 'i3.4xlarge'
+    mixed_scylla:
+        db_type: mixed_scylla
+        instance_type_db: 'i3.large'
+        instance_type_db_oracle: 'i3.8xlarge'
```
Comment (bentsi) [inline tests/gemini_test.yaml]: why `8xlarge?

File: sdcm/tester.py [hunk 21]
```diff
@@ -466,13 +466,24 @@ def create_cluster(db_type='scylla'):
...
+                                      user_prefix=user_prefix + '-oracle'))
```
Comment (bentsi) [inline sdcm/tester.py]: this is hack. need to use class attributes instead of manipulating strings

File: tests/gemini_test.yaml (hunk 19 above)
[the remark this answers is not in the document]
Comment (larisau) [reply tests/gemini_test.yaml]: Gemini uses this file for schema creation, it should be downloaded alongside the binary.

File: sdcm/tester.py [hunk 22]
```diff
@@ -711,6 +722,16 @@ def run_stress_thread_bench(self, stress_cmd, duration=None, stats_aggregate_cmd
                                               self.outputdir,
                                               node_list=self.db_cluster.nodes)
 
+    @clean_resources_on_exception
```
Comment (bentsi) [inline sdcm/tester.py]: we don't need this decorator

File: sdcm/tester.py [hunk 23]
```diff
@@ -748,6 +769,11 @@ def get_stress_results_bench(self, queue):
             self.update_stress_results(results)
         return results
 
+    @clean_resources_on_exception
```
Comment (bentsi) [inline sdcm/tester.py]: don't need this decorator

File: sdcm/cluster.py [hunk 24]
```diff
@@ -2175,33 +2184,38 @@ def node_setup(self, node, verbose=False, db_node_address=None, **kwargs):
...
-            node.remoter.run('sudo yum install git -y')
-            node.remoter.run('curl -LO https://storage.googleapis.com/golang/go1.9.linux-amd64.tar.gz')
-            node.remoter.run('sudo tar -C /usr/local -xvzf go1.9.linux-amd64.tar.gz')
-            node.remoter.run("echo 'export GOPATH=$HOME/go' >> $HOME/.bashrc")
-            node.remoter.run("echo 'export PATH=$PATH:/usr/local/go/bin' >> $HOME/.bashrc")
-            node.remoter.run("source $HOME/.bashrc")
-            node.remoter.run("go get github.com/scylladb/scylla-bench")
         self.get_cassandra_stress_version(node)
 
+        # scylla-bench
```
Comment (bentsi) [inline sdcm/cluster.py]: why did you remove the if?

File: sdcm/cluster.py [hunk 25]
```diff
@@ -2175,33 +2184,38 @@ def node_setup(self, node, verbose=False, db_node_address=None, **kwargs):
...
+        node.remoter.run('sudo yum install git -y')
+        node.remoter.run('curl -LO https://storage.googleapis.com/golang/go1.9.2.linux-amd64.tar.gz')
+        node.remoter.run('sudo tar -C /usr/local -xvzf go1.9.2.linux-amd64.tar.gz')
+        node.remoter.run("echo 'export GOPATH=$HOME/go' >> $HOME/.bashrc")
+        node.remoter.run("echo 'export PATH=$PATH:/usr/local/go/bin' >> $HOME/.bashrc")
+        node.remoter.run("source $HOME/.bashrc")
+        node.remoter.run("go get github.com/scylladb/scylla-bench")
+
+        # gemini tool
+        node.remoter.run("cd $HOME")
```
Comment (bentsi) [inline sdcm/cluster.py]: gemini shouldn't be installed if not needed

File: sdcm/cluster.py [hunk 26]
```diff
@@ -2566,6 +2599,62 @@ def kill_stress_thread_bench(self):
...
+        for node, result in results:
+            output = result.stdout + result.stderr
+            lines = output.splitlines()
+            res = self._parse_gemini_summary(lines)
+            if res:
+                ret.append(res)
+
+        return ret
+
+    def kill_gemini_thread(self):
```
Comment (bentsi) [inline sdcm/cluster.py]: I don't find where this method is called from...

File: sdcm/cluster.py [hunk 27]
```diff
@@ -2175,33 +2184,38 @@ def node_setup(self, node, verbose=False, db_node_address=None, **kwargs):
...
+        node.remoter.run('sudo tar -C /usr/local -xvzf go1.9.2.linux-amd64.tar.gz')
+        node.remoter.run("echo 'export GOPATH=$HOME/go' >> $HOME/.bashrc")
+        node.remoter.run("echo 'export PATH=$PATH:/usr/local/go/bin' >> $HOME/.bashrc")
+        node.remoter.run("source $HOME/.bashrc")
+        node.remoter.run("go get github.com/scylladb/scylla-bench")
+
+        # gemini tool
+        node.remoter.run("cd $HOME")
+        gemini_url = self.params.get('gemini_url')
+        node.remoter.run("curl -LO {}".format(gemini_url))
```
Comment (bentsi) [inline sdcm/cluster.py]: cleaner code:
```
            install_gemini_script = dedent("""
                cd /usr/bin
                curl -L -o gemini {gemini_url}
                chmod a+x gemini
                curl -L0  {gemini_static_url}                
            """.format(**locals()))
            self.remoter.run("bash -cxe '%s'" % install_gemini_script)
```
and you  can run gemini from any user

File: sdcm/cluster.py (hunk 25 above)
Comment (larisau) [reply sdcm/cluster.py]: As I remember it was agreed with @roydahan that we install it with no conditions

File: sdcm/cluster.py (hunk 24 above)
Comment (larisau) [reply sdcm/cluster.py]: see my comment below

File: sdcm/tester.py (hunk 22 above)
Comment (larisau) [reply sdcm/tester.py]: why it still used for cs-stress methods?

Comment (bentsi) [reply sdcm/tester.py]: legacy code, should be removed too

File: sdcm/cluster.py (hunk 25 above)
Comment (bentsi) [reply sdcm/cluster.py]: @roydahan why do we need to install tools that are not used in the test?

File: sdcm/tester.py (hunk 22 above)
Comment (larisau) [reply sdcm/tester.py]: I think you should handle it separately and remove all the usages of this decorator

**Comment by roydahan** [comment]
We will merge it for now, all these changes will be handled as issues / refactor tickets.
